### DISCUSSION TITLE
Descendant of TKGrid

### DISCUSSION THREAD

**Discussion by stefanstefanovs** [kryslt/KControls#64] [General]
I'm trying to register a component as TKGrid descendant in a separate package. It is absolutely empty:
type
  TbasicDataGrid = class(TKCustomControl)
  end;
The package contains only one unit and in the unit is this class definition and the register procedure.
It compiles ok, but when try to install the package get an error "A class named TKCustomControl already exists.". The same thing happens with any other component from this package. I have developed tens of components and never happened something like that. Does someone else experienced a similar issue?

**Comment by kryslt** [comment]
This might be path issue or some old compiled stuff (try Recompile clean or similar).
Btw. 
You should use TKCustomGrid and not TKCustomControl if you want a TKGrid based component:
```
type
TbasicDataGrid = class(TKCustomGrid)
end;
```

**Comment by stefanstefanovs** [comment]
I used TKCustomControl just to test if the same issue appears.
I'm "fighting" with this issue for 2 years.
The final try (today) is on clean Delphi 11.3 installation.
Btw, I get the same error "A class named TKCustomControl already exists." without matter what component I try to descent.

**Comment by kryslt** [comment]
I can confirm this behavior. 
This must be a bug in Delphi, all versions I use (XE and higher) show this error.
When trying in Lazarus the package normally works (see attachment).
I recommend you file a bug report via the Embarcadero Quality Portal.
[package.zip](https://github.com/user-attachments/files/17276037/package.zip)

**Comment by kryslt** [comment]
> This must be a bug in Delphi

Ah sorry, not a bug.
I forgot the weak packaging directive.
https://docwiki.embarcadero.com/RADStudio/Athens/en/Using_the_Weak_Packaging_Directive

  ```
unit DataGrid;
  
  {$WEAKPACKAGEUNIT ON}
  
  interface
  
  uses
    KGrids;
  
  type
    TDataGrid = class(TKCustomGrid)
    //...
    end;
  
  procedure Register;
  
  implementation
  
  uses
    Classes;
  
  procedure Register;
  begin
    RegisterComponents('Tmp', [
      TDataGrid
    ]);
  end;
  
  end.
```

**Comment by stefanstefanovs** [comment]
Thank you for your efforts :)

Unfortunately sound so good to be true...
Yes, with this directive, the package installs, BUT the component is not registered and can not be used in Design. It looks like the Register procedure is not run.

More observations:
- declare the component in a separate unit and add it to run-time package of KControls;
- create an empty unit with only registration procedure and add it into design-time package of KControls;
result with same error!

Repeat the same thing, but do not add the first unit into run-time package. When compile the design-time, the first unit will be "implicitly imported" and the component registers.

Very rare situation for me - looks like a bug in Delphi provoked by a setting or something else into the run-time package.

**Comment by kryslt** [comment]
Hmm, hmm, I can confirm this. 
The new component does not register that way.
Looks like the unit which contains the Register procedure cannot be 'weak' unit (with $WEAKPACKAGEUNIT ON). 
I assume when it is weak then Delphi just ignores it when in conflict and does not call the Register procedure. 

It looks like a Delphi bug indeed to me because when I uncheck the KControls package (KControls Development Suite - design time package) in Components/Install Packages dialog, then my new package (DataGridPkgDesign - see attachment) registers. 
Then, with DataGridPkgDesign registered, the reverse situation occurs and the original KControls package cannot be registered!

Here is what I've found: https://stackoverflow.com/questions/14200789/efilererror-exception-in-delphi-7
This is exactly the issue but no answer (including my own last answer about the WEAKPACKAGEUNIT directive) actually solves this problem.

It is more or less obvious that:
A. Standard VCL controls can be inherited and used in many design time packages installed at the same time. 
B. Third party controls may only be inherited and used in one package at the same time.

[package.zip](https://github.com/user-attachments/files/17284412/package.zip)
